File: maximum-subarray/maximum-subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        //two pointers algorithm

        int i=0;
        int j=0;

        int max_sum=INT_MIN;
        int curr_sum=0;

        while(j<nums.size())
        {
            curr_sum+=nums[j];
            max_sum=max(max_sum,curr_sum);

            if(curr_sum<0)
            {
                curr_sum=0;
                i=j+1;
            }

            j++;
        }

        return max_sum;
        
    }
};

File: 53-maximum-subarray/maximum-subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        int maxsum = Integer.MIN_VALUE;
        int maxtillnow = 0;
        for(int i=0;i<nums.length;i++)
        {
            maxtillnow+=nums[i];
            maxsum = Math.max(maxsum,maxtillnow);
            if(maxtillnow<0)
            {
                maxtillnow = 0;
            }
        }
        return maxsum;
    }
}